pulsar-broker/src/main/java/org/apache/pulsar/broker/delayed/bucket/CombinedSegmentDelayedIndexQueue.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.broker.delayed.bucket;
+
+import java.util.List;
+import javax.annotation.concurrent.NotThreadSafe;
+import org.apache.pulsar.broker.delayed.proto.DelayedMessageIndexBucketSnapshotFormat.DelayedIndex;
+import org.apache.pulsar.broker.delayed.proto.DelayedMessageIndexBucketSnapshotFormat.SnapshotSegment;
+
+@NotThreadSafe
+public class CombinedSegmentDelayedIndexQueue implements DelayedIndexQueue {
+
+    private final List<SnapshotSegment> segmentListA;
+    private final List<SnapshotSegment> segmentListB;
+
+    private int segmentListACursor = 0;
+    private int segmentListBCursor = 0;
+    private int segmentACursor = 0;
+    private int segmentBCursor = 0;
+
+    private CombinedSegmentDelayedIndexQueue(List<SnapshotSegment> segmentListA,
+                                             List<SnapshotSegment> segmentListB) {
+        this.segmentListA = segmentListA;
+        this.segmentListB = segmentListB;
+    }
+
+    public static CombinedSegmentDelayedIndexQueue wrap(
+            List<SnapshotSegment> segmentListA,
+            List<SnapshotSegment> segmentListB) {
+        return new CombinedSegmentDelayedIndexQueue(segmentListA, segmentListB);
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return segmentListACursor >= segmentListA.size() && segmentListBCursor >= segmentListB.size();
+    }
+
+    @Override
+    public DelayedIndex peek() {
+        return getValue(false);
+    }
+
+    @Override
+    public DelayedIndex pop() {
+        return getValue(true);
+    }
+
+    private DelayedIndex getValue(boolean needAdvanceCursor) {
+        while (segmentListACursor < segmentListA.size()
+                && segmentACursor >= segmentListA.get(segmentListACursor).getIndexesCount()) {
+            segmentListACursor++;
+        }
+        while (segmentListBCursor < segmentListB.size()
+                && segmentBCursor >= segmentListB.get(segmentListBCursor).getIndexesCount()) {
+            segmentListBCursor++;
+        }
```

Comment:
It seems a little strange to use `segmentACursor` compare with different segment list sizes, normally, the `segmentACursor` should belong to a specific segment, do these checks are necessary?

Reply:
This is for skipping empty segments.

Reply:
Sorry, I'm not clear about this. If we want to skip the empty segments, could we check whether the index count is zero or not?

Reply:
> Sorry, I'm not clear about this. If we want to skip the empty segments, could we check whether the index count is zero or not?

You are right, This logic is a little redundant, should only check whether the index count is zero or not.